wasm-chain-integration/src/v1/mod.rs

```diff
@@ -0,0 +1,845 @@
+#[cfg(feature = "enable-ffi")]
+mod ffi;
+mod types;
+
+use crate::{constants, v0, ExecResult, InterpreterEnergy, OutOfEnergy};
+use anyhow::{bail, ensure};
+use concordium_contracts_common::{AccountAddress, Amount, ContractAddress, OwnedEntrypointName};
+use machine::Value;
+use std::{borrow::Borrow, io::Write, sync::Arc};
+pub use types::*;
+use wasm_transform::{
+    artifact::{Artifact, CompiledFunction, CompiledFunctionBytes, RunnableCode},
+    machine::{self, ExecutionOutcome, NoInterrupt},
+    utils,
+};
+
+/// Interrupt triggered by the smart contract to execute an instruction on the
+/// host, either an account transfer, or a smart contract.
+#[derive(Debug)]
+pub enum Interrupt {
+    Transfer {
+        to:     AccountAddress,
+        amount: Amount,
+    },
+    Call {
+        address:   ContractAddress,
+        parameter: ParameterVec,
+        name:      OwnedEntrypointName,
+        amount:    Amount,
+    },
+}
+
+impl Interrupt {
+    pub fn to_bytes(&self, out: &mut Vec<u8>) -> anyhow::Result<()> {
+        match self {
+            Interrupt::Transfer {
+                to,
+                amount,
+            } => {
+                out.push(0u8);
+                out.write_all(to.as_ref())?;
+                out.write_all(&amount.micro_ccd.to_be_bytes())?;
+                Ok(())
+            }
+            Interrupt::Call {
+                address,
+                parameter,
+                name,
+                amount,
+            } => {
+                out.push(1u8);
+                out.write_all(&address.index.to_be_bytes())?;
+                out.write_all(&address.subindex.to_be_bytes())?;
+                out.write_all(&(parameter.len() as u16).to_be_bytes())?;
+                out.write_all(&parameter)?;
+                let name_str: &str = name.as_entrypoint_name().into();
+                out.write_all(&(name_str.as_bytes().len() as u16).to_be_bytes())?;
+                out.write_all(&name_str.as_bytes())?;
+                out.write_all(&amount.micro_ccd.to_be_bytes())?;
+                Ok(())
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct InitHost<ParamType, Ctx> {
+    /// Remaining energy for execution.
+    pub energy:            InterpreterEnergy,
+    /// Remaining amount of activation frames.
+    /// In other words, how many more functions can we call in a nested way.
+    pub activation_frames: u32,
+    /// Logs produced during execution.
+    pub logs:              v0::Logs,
+    /// The contract's state.
+    pub state:             v0::State,
+    /// The response from the call.
+    pub return_value:      ReturnValue,
+    /// The parameter to the init method, as well as any responses from
+    /// calls to other contracts during execution.
```

Comment:
This will always just be the single parameter, right? The comment should reflect that.

Reply:
I changed it to a single value now. The previous version allowed interrupts also in init calls, but this is not allowed anymore for simplicity.